src/compiler/moduleNameResolver.ts

```
@@ -96,48 +105,103 @@ namespace ts {
+        const typesVersions = readPackageJsonTypesVersionsField(jsonContent, state);
+        if (typesVersions === undefined) return;
+
+        const result = getPackageJsonTypesVersionsPaths(typesVersions);
+        if (!result) {
+            if (state.traceEnabled) {
+                trace(state.host, Diagnostics.package_json_does_not_have_a_typesVersions_entry_that_matches_version_0, versionMajorMinor);
             }
+            return;
+        }
 
-            const path = normalizePath(combinePaths(baseDirectory, fileName));
+        const { version: bestVersionKey, paths: bestVersionPaths } = result;
+        if (typeof bestVersionPaths !== "object") {
             if (state.traceEnabled) {
-                trace(state.host, Diagnostics.package_json_has_0_field_1_that_references_2, fieldName, fileName, path);
+                trace(state.host, Diagnostics.Expected_type_of_0_field_in_package_json_to_be_1_got_2, `typesVersions['${bestVersionKey}']`, "object", typeof bestVersionPaths);
             }
-            return path;
+            return;
         }
+
+        return result;
     }
 
+    let typeScriptVersion: Version | undefined;
+
     /* @internal */
-    export function readJson(path: string, host: { readFile(fileName: string): string | undefined }): object {
-        try {
-            const jsonText = host.readFile(path);
-            if (!jsonText) return {};
-            const result = parseConfigFileTextToJson(path, jsonText);
-            if (result.error) {
-                return {};
+    export function getPackageJsonTypesVersionsPaths(typesVersions: MapLike<MapLike<string[]>>) {
+        if (!typeScriptVersion) typeScriptVersion = new Version(version);
+
+        for (const key in typesVersions) {
+            if (!hasProperty(typesVersions, key)) continue;
+
+            const keyRange = VersionRange.tryParse(key);
+            if (keyRange === undefined) {

```

This case probably deserves a log message since it's basically a silent error